src/backend/lib/upgrade-storage.js

```diff
@@ -1,271 +1,215 @@
       } else {
-        log(reqId, "createReadStream: asked for hash we don't have:", hash);
-        t.emit("error", new Error(`no such upgrade option with hash ${hash}`));
+        log(`Read finished: ${hash.slice(0, 7)}`);
       }
     });
-    return readonly(t);
-  }
 
-  // Callback<Void> -> Void
-  clearOldApks(cb) {
-    this.localUpgrades.get((err, options) => {
-      if (err) return cb(err);
-      const toDelete = options.filter(
-        option => !this.isNewerThanCurrentApk(option)
-      );
-      const toKeep = options.filter(option =>
-        this.isNewerThanCurrentApk(option)
-      );
-      foreachAsync(
-        toDelete,
-        (option, cb) => {
-          rimraf(option.filename, cb);
-        },
-        _ => {
-          // XXX: not handling errors from failed deletions (yet)
-          this.localUpgrades.set(toKeep, cb);
-        }
-      );
-    });
+    return str;
   }
 
-  // UpgradeOption -> Bool
-  isNewerThanCurrentApk(upgrade) {
-    if (upgrade.arch.indexOf(this.targetArch) === -1) return null;
-    if (upgrade.platform !== this.targetPlatform) return null;
-    if (!semver.valid(upgrade.version)) return null;
-    return semver.gt(upgrade.version, this.version);
-  }
-
-  apkToUpgradeOption(filepath, version, cb) {
-    hashFile(filepath, (err, buf) => {
-      if (err) return cb(err);
-      const hash = buf.toString("hex");
-      fs.stat(filepath, (err, info) => {
-        if (err) return cb(err);
-        const option = {
-          filename: filepath,
-          hash,
-          size: info.size,
-          version,
-          hashType: "sha256",
-          platform: "android",
-          arch: ["arm64-v8a"],
-          id: hash,
-        };
-        cb(null, option);
-      });
+  /**
+   * Create a writable stream for adding an installer to storage.
+   *
+   * @param {object} options
+   * @param {string} options.hash Expected sha256 hash of the incoming file
+   */
+  createWriteStream({ hash: expectedHash }) {
+    const tmpFilepath = path.join(
+      this.#tmpdir,
+      crypto.randomBytes(8).toString("hex") + ".apk"
+    );
+    return beforeAfterStream({
+      createStream: async () => {
+        log(`Write started: ${expectedHash.slice(0, 7)}`);
+        await this.started();
+        return fs.createWriteStream(tmpFilepath);
+      },
+      finalize: async fsStream => {
+        try {
+          // Everything after here happens after createWriteStream has finished, but
+          // that's ok because this is UpgradeStorage updating its state
+          const installer = await getInstallerInfo(tmpFilepath);
+          if (!installer) {
+            // We couldn't read the info about this installer
+            throw new Error("Could not parse APK");
+          }
+          if (installer.hash !== expectedHash) {
+            // Hashes did not match, download was broken or incorrectly advertised
+            throw new Error("Invalid hash");
```

Comment:
```suggestion
            throw new Error("Actual hash does not match expected hash");
```
The hash is [likely] a valid sha256 hash, but it's just not the value we expected it to be.

Reply:
True! Thanks.